module/zcommon/zfeature_common.c

```diff
@@ -348,6 +348,19 @@ zpool_feature_init(void)
 	    ZFEATURE_FLAG_MOS | ZFEATURE_FLAG_ACTIVATE_ON_ENABLE,
 	    ZFEATURE_TYPE_BOOLEAN, NULL);
 
+	{
+	static const spa_feature_t log_spacemap_deps[] = {
+		SPA_FEATURE_SPACEMAP_V2,
+		SPA_FEATURE_NONE
+	};
+	zfeature_register(SPA_FEATURE_LOG_SPACEMAP,
+	    "com.delphix:log_spacemap", "log_spacemap",
+	    "Log metaslab changes on a single spacemap and "
+	    "flush them periodically.",
+	    ZFEATURE_FLAG_READONLY_COMPAT, ZFEATURE_TYPE_BOOLEAN,
```

Comment:
Have we tried doing a readonly open of a pool with log spacemaps with older software?  I know that in principle we don't need to look at the spacemaps to do a readonly open, but I vaguely recall that older software tried to load metaslabs regardless, which might cause problems?  Maybe do a spot check of moving a pool with log spacemaps back to 0.7.x (readonly).

Reply:
I remember testing this illumos but I redid a quick experiment in Linux.
I created 2 VMs -
[A] latest bits but without the Log Spacemap feature
[B] latest bits patched with the Log Spacemap feature.

The I did the following steps:
1) I created a file-based pool in [B] that has the feature activated:
```
$ truncate -s 512m dsk
$ sudo zpool create testpool $(pwd)/dsk
$ zpool list
NAME       SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
rpool     69.5G  14.1G  55.4G        -         -      -    20%  1.00x    ONLINE  -
testpool   480M   129K   480M        -         -     0%     0%  1.00x    ONLINE  -
$ sudo zpool get feature@log_spacemap testpool
NAME      PROPERTY              VALUE                 SOURCE
testpool  feature@log_spacemap  active                local
``` 
2) Then exported the pool and ensured that there were at least a few spacemap logs (not that it matters since the feature is on regardless. I was also running debug bits which randomly decided to not clean spacemap logs during export):
```
$ sudo zpool export testpool
$ sudo zdb -m -e -p . testpool
....
Log Space Maps in Pool:
Log Spacemap object 65 txg 4
space map object 65:
  smp_length = 0x1e0
  smp_alloc = 0x16200
Log Spacemap object 76 txg 5
space map object 76:
  smp_length = 0x3a0
  smp_alloc = 0x2400
Log Spacemap object 77 txg 6
space map object 77:
  smp_length = 0x268
  smp_alloc = 0x1e00
Log Spacemap object 78 txg 17
space map object 78:
  smp_length = 0x208
  smp_alloc = 0x1800
Log Spacemap object 79 txg 26
space map object 79:
  smp_length = 0x2e0
  smp_alloc = 0x1e00

Log Space Map Obsolete Entry Statistics:
10       valid entries out of 20       - txg 4
30       valid entries out of 43       - txg 5
24       valid entries out of 28       - txg 6
20       valid entries out of 22       - txg 17
33       valid entries out of 33       - txg 26
117      valid entries out of 146      - total
```
3) Copied the pool over to [A] and tried to import it
```
$ scp <host B>:~/dsk .
dsk                                                                                                   100%  512MB 144.2MB/s   00:03
$ sudo zpool import -d . testpool
This pool uses the following feature(s) not supported by this system:
	com.delphix:log_spacemap (Log metaslab changes on a single spacemap and flush them periodically.)
All unsupported features are only required for writing to the pool.
The pool can be imported using '-o readonly=on'.
cannot import 'testpool': unsupported version or feature
```
4) Import the pool read-only:
```
$ sudo zpool import -o readonly=on -d . testpool
$ zpool list
NAME       SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
rpool     69.5G  13.8G  55.7G        -         -      -    19%  1.00x    ONLINE  -
testpool   480M    36K   480M        -         -     0%     0%  1.00x    ONLINE  -
$ sudo zpool get all testpool
....
testpool  unsupported@com.delphix:log_spacemap  readonly                              local
```

Does this testing cover your question? If yes I'll also add it in the PR description.

Reply:
Looks good, thanks.